runtime/near-vm-logic/src/types.rs

```diff
@@ -50,5 +52,156 @@ pub enum PromiseResult {
     Failed,
 }
 
+const PROFILE_DATA_LEN: usize = 1 + ActionCosts::count() + ExtCosts::count();
+
+/// Data for total gas burnt (index 0), and then each action and external cost
+pub type InternalProfileData<T> = [T; PROFILE_DATA_LEN];
+
+pub struct FixedArray<T> {
+    pub data: Rc<RefCell<InternalProfileData<T>>>
+}
+
+
+impl<T> Clone for FixedArray<T> {
+    fn clone(&self) -> Self {
+        Self {
+            data: Rc::clone(&self.data)
+        }
+    }
+}
+
+
 /// Profile of gas consumption.
-pub type ProfileData = Rc<RefCell<[u64; ActionCosts::count() + ExtCosts::count()]>>;
+pub type ProfileData = FixedArray<u64>;
+
+pub fn clone_profile(profile: &ProfileData) -> ProfileData {
+  profile.clone()
+}
+
+impl ProfileData {
+
+  const EXT_START: usize = 1;
+  const ACTION_START: usize = ProfileData::EXT_START + ExtCosts::count();
+  // const LENGTH: usize = PROFILE_DATA_LEN;
+  pub fn all_gas(&self) -> Gas {
+    (*self.data.borrow())[0]
+  }
+
+  fn borrow(&self) -> InternalProfileData<u64> {
+    *self.data.borrow()
+  }
+
+  fn add_val(&self, index: usize, value: u64) {
+    *self.data.borrow_mut().get_mut(index).unwrap() += value;
+  }
+
+  pub fn add_action_cost(&self, action: ActionCosts, value: u64) {
+    self.add_val(ProfileData:: ACTION_START + action as usize, value);
```

Comment:
should we add to both action_cost and all_gas here? Also add_ext_cost below

Reply:
It is set after execution as gas costs can be from other sources than these two.